marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochState.hs

```diff
@@ -591,55 +591,69 @@ instance Rewindable EpochStateHandle where
-              Nothing                              -> pure ()
-              Just (_, slotNo, _, _) | slotNo > sn -> removeFile $ ledgerStateDirPath </> fp
-              Just _                               -> pure ()
+            case readLedgerStateFileMetadata fp of
+              Nothing                           -> pure ()
+              Just lsm | lsfMetaSlotNo lsm > sn -> removeFile $ ledgerStateDirPath </> fp
+              Just _                            -> pure ()
 
         pure h
 
 instance Resumable EpochStateHandle where
     resumeFromStorage
         :: EpochStateHandle
         -> StorableMonad EpochStateHandle C.ChainPoint
-    resumeFromStorage (EpochStateHandle _ c ledgerStateDirPath _)
-        = liftSQLError CantQueryIndexer $ do
-        ledgerStateFilepaths <- listDirectory ledgerStateDirPath
-        let ledgerStateChainPoints =
-                fmap (\(_, sn, bhh, _) -> (sn, bhh))
-                $ mapMaybe chainTipsFromLedgerStateFilePath ledgerStateFilepaths
-
-        epochSDDChainPoints <- flip filterM ledgerStateChainPoints $ \(slotNo, _) -> do
-            result :: [[C.SlotNo]] <- SQL.query c
-                [r|SELECT slotNo
-                   FROM epoch_sdd
-                   WHERE slotNo = ? LIMIT 1 |] (SQL.Only slotNo)
-            pure $ not $ null result
-
-        epochNonceChainPoints <- flip filterM ledgerStateChainPoints $ \(slotNo, _) -> do
-            result :: [[C.SlotNo]] <- SQL.query c
-                [r|SELECT slotNo
-                   FROM epoch_nonce
-                   WHERE slotNo = ? LIMIT 1 |] (SQL.Only slotNo)
-            pure $ not $ null result
-
-        let resumablePoints = List.sortOn Down
-                            $ fmap (uncurry C.ChainPoint)
-                            $ Set.toList
-                            $ Set.intersection (Set.fromList epochSDDChainPoints)
-                                               (Set.fromList epochNonceChainPoints)
-        pure $ chainPointOrGenesis resumablePoints
-
-chainTipsFromLedgerStateFilePath :: FilePath -> Maybe (Bool, C.SlotNo, C.Hash C.BlockHeader, C.BlockNo)
-chainTipsFromLedgerStateFilePath ledgerStateFilepath =
+    resumeFromStorage (EpochStateHandle topLevelConfig c ledgerStateDirPath _) =
+        liftSQLError CantQueryIndexer $ do
+            -- We first identify the chain points from the LedgerState files that we stored on disk.
+            -- We ignore volatile ledger state files and we try to deserialise the LedgerState to
+            -- ensure that we correctly serialised it in a previous run. If the file is not
+            -- deserializable, we delete it.
+            ledgerStateFilepaths <- listDirectory ledgerStateDirPath
+            readableLedgerStateFilePaths <- flip filterM ledgerStateFilepaths $ \ledgerStateFilePath -> do
+                lsM <- readLedgerStateFromDisk (ledgerStateDirPath </> ledgerStateFilePath) topLevelConfig
+                case lsM of
+                  Nothing -> pure False
```

Comment:
Maybe delete the file already here? It won't be any use later on anyway..

Reply:
The deletion logic is already implemented in `rewindStorage`. Once we start the chain-sync protocol with the point we're going to resume from, a `RollBackward` event is emitted, which in our case will delete the corrupt LedgerState given the implement of `rewindStorage`. I thought that I wouldn't want to duplicate the logic here. Does that make sense to you?

Reply:
Isn't there a danger that whenever ledger states get removed then accidentally a corrupt one gets kept? Maybe the system as a whole functions in a way that this doesn't occur, but it seems like if nothing bad can happen when the file is deleted right away then just delete it right away.

Reply:
See 479869bf8868db054f4c794b413eea492465fbaa